File: ServiceTemplates/Basic/HTTP_SERVER.py
import asyncio
from fastapi import FastAPI
import uvicorn




# This is Template for a Class that has the Base HTTP class as an object and All the Routes and end points for the HTTP Server Are Defined Here
class MainServer:
    def __init__(self, httpServerHost, httpServerPort):
        self.fast = HTTPServer(httpServerHost, httpServerPort)

    # Define the Routes for the Particular Case
    def ConfigureServerRoutes(self):

        @self.fast.app.get("/")
        async def read_root():
            print("Running Through Someone Else")
            return {"message": "Hello World"}
    
    async def RunServer(self):
        self.ConfigureServerRoutes()
        await self.fast.run_app()


# This is the Base Class Which can be Called to Setup a HTTP Server
class HTTPServer:
    def __init__(self, host="127.0.0.1", port=54545):
        self.app = FastAPI()
        self.host = host
        self.port = port

    async def run_app(self):
        config = uvicorn.Config(self.app, host=self.host, port=self.port)
        server = uvicorn.Server(config)
        await server.serve()


async def start_server():
    server = MainServer('127.0.0.1', 8000)
    await server.RunServer()
    pass

# if __name__ == "__main__":
#     import asyncio
#     asyncio.run(start_server())

File: ServiceTemplates/Basic/HTTP_SERVER.js
const express = require('express');

class HTTPServer {
  constructor(host = "127.0.0.1", port = 54545) {
    this.app = express();
    this.host = host;
    this.port = port;
  }

  async run_app() {
    // Wrap the listen method in a promise so it behaves asynchronously.
    return new Promise((resolve, reject) => {
      this.server = this.app.listen(this.port, this.host, () => {
        console.log(`Server running on http://${this.host}:${this.port}`);
        resolve();
      });
      // Optional: handle server errors
      this.server.on('error', (err) => {
        console.error("Server error:", err);
        reject(err);
      });
    });
  }
}

class MainServer {
  constructor(httpServerHost, httpServerPort) {
    this.fast = new HTTPServer(httpServerHost, httpServerPort);
  }

  // Define the routes similar to FastAPI route decorators.
  ConfigureServerRoutes() {
    this.fast.app.get("/", async (req, res) => {
      console.log("Running Through Someone Else");
      res.json({ message: "Hello World" });
    });
  }

  async RunServer() {
    this.ConfigureServerRoutes();
    await this.fast.run_app();
  }
}

// Async function to start the server.
async function start_server() {
  const server = new MainServer('127.0.0.1', 8000);
  await server.RunServer();
}

// Start the server.
start_server().catch(err => {
  console.error("Failed to start server:", err);
});
